web/src/stream_color.js

```diff
@@ -1,42 +1,71 @@
+    if ($stream_header.length) {
+        const stream_id = Number.parseInt($($stream_header).attr("data-stream-id"), 10);
+        const stream_name = stream_data.maybe_get_stream_name(stream_id);
+        if (!stream_name) {
+            return;
+        }
+        const stream_color = stream_data.get_color(stream_name);
+        const recipient_bar_color = get_recipient_bar_color(stream_color);
+        $stream_header
+            .find(".message-header-contents")
+            .css("background-color", recipient_bar_color);
+    }
+}
+
+export function get_stream_privacy_icon_color(color) {
+    // LCH stands for Lightness, Chroma, and Hue.
+    // This function restricts Lightness of a color to be between 20 to 75.
+    color = colord(color).toLch();
+    const min_color_l = 20;
+    const max_color_l = 75;
+    if (color.l < min_color_l) {
+        color.l = min_color_l;
+    } else if (color.l > max_color_l) {
+        color.l = max_color_l;
+    }
+    return colord(color).toHex();
+}
+
+export function get_recipient_bar_color(color) {
+    // Mixes 20% of color to 80% of white (light theme) / black (dark theme).
+    const using_dark_theme = settings_data.using_dark_theme();
+    color = get_stream_privacy_icon_color(color);
+    return colord(using_dark_theme ? "#000000" : "#ffffff")
+        .mix(color, using_dark_theme ? 0.5 : 0.4)
+        .toHex();
+}
 
+function update_table_stream_color(table, stream_name, color) {
+    const recipient_bar_color = get_recipient_bar_color(color);
+    const $stream_labels = table.find(".stream_label");
     for (const label of $stream_labels) {
         const $label = $(label);
         if ($label.text().trim() === stream_name) {
-            const $messages = $label.closest(".recipient_row").children(".message_row");
-            $messages
-                .children(".messagebox")
-                .css(
-                    "box-shadow",
-                    "inset 2px 0px 0px 0px " + style + ", -1px 0px 0px 0px " + style,
-                );
-            $messages
-                .children(".date_row")
-                .css(
-                    "box-shadow",
-                    "inset 2px 0px 0px 0px " + style + ", -1px 0px 0px 0px " + style,
-                );
-            $label.css({background: style, "border-left-color": style});
-            $label.removeClass("dark_background");
-            $label.addClass(color_class.get_css_class(color));
+            $label
+                .closest(".message_header_stream .message-header-contents")
+                .css({background: recipient_bar_color});
         }
     }
 }
 
 function update_stream_privacy_color(id, color) {
-    $(`.stream-privacy-${CSS.escape(id)}`).css("color", color);
+    $(`.stream-privacy-${CSS.escape(id)}.original-color`).css("color", color);
+    color = get_stream_privacy_icon_color(color);
+    $(`.stream-privacy-${CSS.escape(id)}.modified-color`).css("color", color);
```

Comment:
We should always prefer class names that are unique by themselves, rather than multiple non-unique class names whose combination happens to be unique—the latter gets really confusing when we’re trying to refactor anything.

There doesn’t seem to be a corresponding update for the recipient bar background color.

Reply:
> There doesn’t seem to be a corresponding update for the recipient bar background color.

They are updated separately by `update_message_recipient_color`.

> We should always prefer class names that are unique by themselves, rather than multiple non-unique class names whose combination happens to be unique—the latter gets really confusing when we’re trying to refactor anything.

Changed!